Sources/FoundationEssentials/Calendar/Calendar_Recurrence.swift

```diff
@@ -0,0 +1,846 @@
+                    }
+                }
+                return nil
+            }
+        }
+        
+        public func makeIterator() -> Iterator {
+            return Iterator(start: start, matching: recurrence, range: range)
+        }
+    }
+}
+
+@available(FoundationPreview 0.4, *)
+extension Calendar.RecurrenceRule {
+    /// Move each date to the given weeks of the year
+    internal func _expandWeeks(dates: inout [Date], anchor: Date) {
+        guard
+          let yearInterval = calendar.dateInterval(of: .year, for: anchor),
+          let weekRange = calendar.range(of: .weekOfYear, in: .year, for: anchor)
+          else {
+            return
+        }
+        
+        /// The weekday on which the first day of the year falls
+        let firstWeekdayOfYear = calendar.component(.weekday, from: yearInterval.start)
+        /// The weekday on which the last day of the year falls. We remove a few
+        /// seconds from the end of the range, since it falls on January 1 00:00
+        /// the following year.
+        let lastWeekayOfYear = calendar.component(.weekday, from: yearInterval.end.addingTimeInterval(-0.01))
+        
+        let minimumDaysInFirstWeek = calendar.minimumDaysInFirstWeek
+        let firstWeekday = calendar.firstWeekday
+        
+        /// How many days of the first week are within the year
+        let daysInFirstWeek = 7 - firstWeekdayOfYear + firstWeekday
+        /// How many days of the last week are within the year
+        let daysLeftInLastWeek = 7 - lastWeekayOfYear + firstWeekday
+        
+        
+        let firstWeekIdx = if daysInFirstWeek >= minimumDaysInFirstWeek {
+            weekRange.lowerBound
+        } else {
+            weekRange.lowerBound + 1
+        }
+        
+        let lastWeekIdx = if daysLeftInLastWeek >= minimumDaysInFirstWeek {
+            weekRange.upperBound - 2
+        } else {
+            weekRange.upperBound - 1
+        }
+        
+        let weeks = weeks.map { weekIdx in
+           if weekIdx > 0 {
+               weekIdx - 1 + firstWeekIdx
+           } else {
+               lastWeekIdx + (weekIdx + 1)
+           }
+        }
+        
+        dates = dates.flatMap { date in
+            let week = calendar.component(.weekOfYear, from: date)
+            return weeks.compactMap { weekIdx in
+                let offset = weekIdx - week
+                return calendar.date(byAdding: .weekOfYear, value: offset, to: date)
+            }
+        }
+    }
+    
+    internal func _expandOrLimitMonths(dates: inout [Date], anchor: Date, action: ComponentAction) {
+        lazy var monthRange = calendar.range(of: .month, in: .year, for: anchor)!
+        let months = months.map { month in
+            if month.index > 0 {
+                return month
+            } else {
+                let newIndex = monthRange.upperBound + month.index
+                // The upper bound is the last month plus one. Subtracting 1 we get the last month
```

Comment:
Is this comment still correct?

Reply:
yeah! the upper bound would be 13 in the Gregorian calendar. So the last month is 13 + (-1) = 12